packages/fast-usdc/src/exos/liquidity-pool.js

```diff
@@ -0,0 +1,185 @@
+ */
+
+import { AmountMath } from '@agoric/ertp/src/amountMath.js';
+import {
+  AmountShape,
+  PaymentShape,
+  RatioShape,
+} from '@agoric/ertp/src/typeGuards.js';
+import { depositToSeat } from '@agoric/zoe/src/contractSupport/zoeHelpers.js';
+import { SeatShape } from '@agoric/zoe/src/typeGuards.js';
+import { M } from '@endo/patterns';
+import {
+  makeRecorderTopic,
+  TopicsRecordShape,
+} from '@agoric/zoe/src/contractSupport/topics.js';
+import {
+  deposit as depositCalc,
+  makeParity,
+  withdraw as withdrawCalc,
+  withFees,
+} from '../pool-share-math.js';
+import { makeProposalShapes } from '../type-guards.js';
+
+/**
+ * @param {Zone} zone
+ * @param {ZCF} zcf
+ * @param {Record<'USDC', Brand<'nat'>>} brands
+ * @param {{
+ *   makeRecorderKit: MakeRecorderKit;
+ * }} tools
+ */
+export const prepareLiquidityPoolKit = (zone, zcf, brands, tools) => {
+  return zone.exoClassKit(
+    'Fast Liquidity Pool',
+    {
+      feeSink: M.interface('feeSink', {
+        receive: M.call(AmountShape, PaymentShape).returns(M.promise()),
+      }),
+      external: M.interface('external', {
+        setShareWorth: M.call(RatioShape).returns(M.promise()),
+      }),
+      depositHandler: M.interface('depositHandler', {
+        handle: M.call(SeatShape, M.any()).returns(M.promise()),
+      }),
+      withdrawHandler: M.interface('withdrawHandler', {
+        handle: M.call(SeatShape, M.any()).returns(M.promise()),
+      }),
+      public: M.interface('public', {
+        makeDepositInvitation: M.call().returns(M.promise()),
+        makeWithdrawInvitation: M.call().returns(M.promise()),
+        getPublicTopics: M.call().returns(TopicsRecordShape),
+      }),
+    },
+    /**
+     * @param {ZCFMint<'nat'>} shareMint
+     * @param {StorageNode} node
+     */
+    (shareMint, node) => {
+      const { brand: PoolShares } = shareMint.getIssuerRecord();
+      const { USDC } = brands;
+      const proposalShapes = makeProposalShapes({ USDC, PoolShares });
+      const dust = AmountMath.make(USDC, 1n);
+      const shareWorth = makeParity(dust, PoolShares);
+      const { zcfSeat: poolSeat } = zcf.makeEmptySeatKit();
+      const shareWorthRecorderKit = tools.makeRecorderKit(
+        node,
+        /** @type {TypedPattern<ShareWorth>} */ (RatioShape),
+      );
+      return {
+        shareMint,
+        shareWorth,
+        poolSeat,
+        PoolShares,
+        proposalShapes,
+        shareWorthRecorderKit,
+      };
+    },
+    {
+      feeSink: {
+        /**
+         * @param {Amount<'nat'>} amount
+         * @param {Payment<'nat'>} payment
+         */
+        async receive(amount, payment) {
+          const { poolSeat, shareWorth } = this.state;
+          const { external } = this.facets;
+          await depositToSeat(
+            zcf,
+            poolSeat,
+            harden({ USDC: amount }),
+            harden({ USDC: payment }),
+          );
+          await external.setShareWorth(withFees(shareWorth, amount));
```

Comment:
this looks like something that could fail and need recovery. In other contracts I think we've modified the state locally (synchronously) and then fired of a "publish" call.

E.g.
```suggestion
          this.state.sharedWorth = shareWorth;
          external.publishShareWorth();
```
It's a little longer but makes it clear how the data is flowing. There's now way to recover from a failed recorder `.write`.

If that doesn't look good to you let's chat higher bandwidth

Reply:
I had `void` instead of `await` originally. Lack of synchronization made my tests go wonky. But I guess that's just life in the big city. The test code will have to try harder to sync up.

As to whether the assignment is inside or out, it seemed useful to be clear that the assignment is _always_ paired with publication. But I don't feel strongly about it. I suppose the failure semantics are more clear with your suggestion, and that's more important.

Reply:
oh. oops. `setShareWorth` being `async` introduces a turn boundary, doesn't it. silly me.